src/Build/BackEnd/Components/RequestBuilder/IntrinsicTasks/ItemGroupLoggingHelper.cs

```diff
@@ -95,53 +93,93 @@ internal static string GetParameterText(string prefix, string parameterName, ILi
         /// </summary>
         internal static string GetStringFromParameterValue(object parameterValue)
         {
-            var type = parameterValue.GetType();
-
-            ErrorUtilities.VerifyThrow(!type.IsArray, "scalars only");
+            // fast path for the common case
+            if (parameterValue is string valueText)
+            {
+                return valueText;
+            }
 
-            if (type == typeof(string))
+            using (var sb = new ReuseableStringBuilder())
             {
-                return (string)parameterValue;
+                AppendStringFromParameterValue(sb, parameterValue);
+                return sb.ToString();
             }
-            else if (type.GetTypeInfo().IsValueType)
+        }
+
+        // Avoid allocating a temporary list to hold metadata for sorting every time.
+        // Each thread gets its own copy.
+        [ThreadStatic]
+        private static List<KeyValuePair<string, string>> keyValuePairList;
+
+        private static void AppendStringFromParameterValue(ReuseableStringBuilder sb, object parameterValue)
```

Comment:
Separate PR, if idea is worth it: I have a hunch this method is called a lot of times with the same `parameterValue` instance. Worth caching?

Reply:
Caching brings both memory cost as well as string hashcode calculation and dictionary lookup. Probably not worth it here and adds complexity.

We'll get much bigger wins if we stream the raw items to the loggers without using a string builder at all.